### PR TITLE
Glimpse draft

### LINKED ISSUE #1007: Feature request: Implement Glimpse Draft
[Specification](http://turnonemagic.com/glimpse-drafting-best-way-draft-cube-2-4-players-title-shorter/) of Glimpse Draft format.

I'll implement it after I (or someone else) will fix #1006

### PR DESCRIPTION
**Pull request by HerveH44**
## Linked tickets
- Fixes #1007
- Supersedes #1021

## Explanation of the issue
TODO

### Autopick / Burn Menu 
How to cancel autopick or burnpick ?
With 1  pick per pack, you just pick something else. With 2 or more, the current solution is that the next pick cancels the oldest autopick (like a fixed queue where the new entry pushed all entries and the oldest one may fall out of the queue). 

My idea is to simplify it to have a contextual view on click on a card :
Autopick / Burn card
This would then be updated if the card is already selected for one of the two:
Cancel autopick / burn card 
Or
Autopick / cancel burn card
@mixmix : What do you think of this behavior ? Would be ready to implement something like this, like an overlay with 2 options (Autopick / Burn) ?

### Pick delegates
This is hard to understand. With glimpse it looks like we can see the common behavior of normal, decadent and glimpse draft using both pick and burn actions: 
Normal: pick 1 and burn 0 
glimpse : pick 1 and burn 2
decadent: pick 1 and burn the rest
In that perspective, we could unify the pickdelegate to one function that pick the cards in autopick, burn the cards in burnpick (for decadent, the idea is to set the burnpicks number to pack's size minus cards to pick (e.g. a pack of 14 cards, with 1 card to pick would have in decadent draft, 14-1=13 cards to burn). 

## Screenshots
TODO

### CODE REVIEW

**Comment by LotusWalker** [comment]
Herve: Thanks so much for your coding efforts to implement the Glimpse Draft format on Dr4ft.info! I’ve been looking forward to this capability for a long time now, and it will be great to have some cube drafts vs. friends without having bots clog up the process. Drop me your paypal and I’ll ship you some $$ in appreciation.

**Comment by HerveH44** [comment]
@LotusWalker you're welcome. I hope I can manage to finish this before christmas if I have enough time. 

@lrawicz @mixmix @tooomm (maybe others but I don't know who is still working on dr4ft): I've noticed that normal, decadent and glimpse draft could be defined with the couple picksPerPack and burnsPerPack like so : 
normal = 1 pickPerPack and 0 burn
decadent = 1 pickPerPack and burn the rest
glimpse = 1 pick 2 burn
If I'm correct, we could simplify the draft picking mechanism to a simple pick x target cards and burn x target cards for all our types of game (no more pickDelegate). Maybe I'm missing something, let me know what you think

I'd like to talk about that and my implementation at some point, would you be available to quick zoom to confront our point of views. When would you be available?

**Comment by ZeldaZach** [comment]
@HerveH44 I'm not around today, but I'm around all week at all hours normally. Just send me a Discord msg. I like your idea of supporting picksPerPack and burnsPerPack. Would simplify the service and allow multiple draft formats to work, just by setting two variables

**Comment by mixmix** [comment]
I think the idea of a context menu could work but we need to be careful to not add heaps of clicks. Actually I don't like the idea that you click to see context, then click to choose pick/ burn. That's 6 clicks per pack (minimum).

My alternative UX would be that with glimpse there are two "rounds" of picking. First you pick your card to take (the behaves as normal). 
Then you pick your burns - this could be a queue where if you select a third card the oldest falls off but I think actually "toggling" cards is something people understand.
To initiate the burn people click a button which says "burn".

I would be keen to unify the design for living so that one click selects, and there's a button which you then click called "pick".
I think we need this because the existing double click behavior is very unintuitive to new players.

I'll draw some rough sketches for different ways this could look.

ITO state management, I'm a big fan of refactors which improve clarity and maintainability. I would strongly encourage you to do an initial piece of work on a separate branch which is exclusively a refactor (no functionality changes). You can do it in a way which sets this next block up well. Don't it in two steps will make it easier to review and merge the changes

**Comment by HerveH44** [comment]
I like your UI idea of 2 rounds but I don't know how we could implement it yet. I am wondering if that should be the responbility of the Frontend or Backend. I would tend to just frontend as touching the backend picking mechanism is always a bit risky.

**Comment by mixmix** [comment]
I've thought about the 2 rounds more and realized I've projected how I play glimpse too much. I don't think we should force puerile to pick before burn. Maybe you know what you're burning first because these some bombs not in your colors.

I'll have a think about logic and draw something up. I'm trying to avoid using mouse-over effects because they're no good on mobile. But maybe they're an OK trade-off

**Comment by LotusWalker** [comment]
Also, while Burn 2 cards is the most common way to play Glimpse, there are variations depending on how many people are drafting where you might want to Burn 1, 3, or 4 cards.

**Comment by ZeldaZach** [comment]
> My alternative UX would be that with glimpse there are two "rounds" of picking. First you pick your card to take (the behaves as normal).
Then you pick your burns - this could be a queue where if you select a third card the oldest falls off but I think actually "toggling" cards is something people understand.
To initiate the burn people click a button which says "burn".

My idea for the UI here would be a green wrapper around the cards you're picking, and a red border around the cards you're burning. I'm not sure the best UX interaction for players, as my initial thought was pick your N cards, then pick your burn M cards. People might change their minds, so it could get tricky. Could also do the multiple click option. Single click = Green, Double click = Red. Clicking on a green makes it red, clicking on a red makes it unselected.

This would be diff from our system rn as a double click will select the cards, if able. Just throwing out ideas.

**Comment by mixmix** [comment]
I thought about red and green too. I'm gonna draw up an alternative with symbols which overlay the image part of the card, as colors don't confirm what they mean so strongly, and red and green have accessibility problems (some sorts of colour blind would eat shit!)

**Comment by HerveH44** [comment]
I would tend to the first solution of @mixmix. I imagine adding a new phase , "burn phase" to the picks. That would not change anything from the drafter, it would be the same game feel, click to select the cards you want (1st phase pick, 2nd phase burn). Everything would be implemented from the backend that sends the pack and tell the frontend in which phase he is in (so we can indicate accurately in frontend if the cards to be selected are going to the deck or to be discarded).

It's contrary to what I've implemented so far (I went for my first solution), but I'm happy to start over because in the end it would be way less change (the frontend would just need the new indicator "cards to burn" and a new argument when you receive the pack which lets you know in which phase you are). The backend would also be very simple. 

so to summarize, I would go with what changes less the actual game feel, how you click to select the cards. I don't think adding a second phase (and thus having to pick first then discard the cards) would be a hard thing to learn from players. It would be harder to handle the contextual menu I'm working on in my current implementation. 

Let me know if I could convince you.

**Comment by mixmix** [comment]
I think I'm really into the single pass (pick + burn), followed by "confirm" to execute those selections (or auto execute if time reached). 

I'm still worried about that problem where a person might like to burn before picking, or they pick and then realize when they start into burning that they want to pick something they were going to burn! But now they're locked in.

I think those problems I describe are only really a problem when there is a time limit and auto picking involved.

I don't play with a time limit so this doesn't effect me. But I'm up for bigger changes which add more clarity (to picking process for app in general) and support more people.

**Comment by mixmix** [comment]
(but I'm not a strong block - I think my suggestion is better in the long run, but I'm not going to die if we go with the two round solution)

**Comment by mixmix** [comment]
Here's ideas for what it might look like when you mouse over a card. This is assuming you are picking and burning at same time



I've used a trash can instead of a 🔥 flame only because it was easy to sketch on phone and because perhaps it's a more universal symbol for "get rid of". Don't mind tho

Here what a card that has been queued up to be picked might look like



In both cases I'm placing those symbols in places which I think are easy to click, or are out of the way for people sucky reviewing their pick (the name of the card is probably the least valuable piece of info?)

**Comment by HerveH44** [comment]
@mixmix it looks nice. Just watch out where you put the overlay because the card might just text (no image) so you might cover the title or some other important informations!

**Comment by mixmix** [comment]
Oh good point about text only mode. I think if we're ok using on-hover to show these buttons, then it will only be a temporary obscuring.

**Comment by tooomm** [comment]
I also think doing everything in one round is way more flexible and a user can review all his selection in context rather than isolation for better evaluation before shipping his choices.

Another idea how to handle picking and clearly marking cards for burning could be two new drop zones you drag the specific cards to before confirming your selection and proceeding. That might not work best with the current layout for drafting/picking. But wanted to drop the concept to the discussion.

**Comment by mixmix** [comment]
Drag and drop requires a whole lot more fiddling I think. I did wonder about auto moving selected cards to a special area. Main concern with that would be that it results in test of cards jumping around

**Comment by HerveH44** [comment]
Hey @mixmix. I've been waiting to get your svgs or frontends helpers. Are you planning on creating some? I've stopped a bit working on this but if you can provide it to me I can then try to wrap up the rest.

**Comment by mixmix** [comment]
Hey sorry got caught up christmas and family stuff. Will do this in the next couple days

**Comment by mixmix** [comment]
Do you want me to code the front end where the svgs are displayed and have some state and onClick which you can then hook into?

If yes cab you point me at a branch and file your like me to weave into?

**Comment by HerveH44** [comment]
Yeah, you have in the Grid.jsx file the PackCardContextualMenuDecorator  class that handles the logic. You can work from there I think

**Comment by HerveH44** [comment]
I think we could simplify even more with the couple picksPerPack and burnPerPack. For example a sealed could be a draft where you pick all the cards of each pack. But it would be done automatically by the game. In the end, we could reuse the same game handler for both draft and sealed. What do you think @ZeldaZach ?

File: backend/util.js [hunk 1]
```diff
@@ -10,14 +10,15 @@ const BASICS = [
...
-  let {list, cards, packs, cubePoolSize} = cube;
+  let {list, cards, packs, cubePoolSize, burnsPerPack} = cube;
 
   assert(typeof list === "string", "cube.list must be a string");
   assert(typeof cards === "number", "cube.cards must be a number");
   assert(typeof cubePoolSize === "number", "cube.cubePoolSize must be a number");
   assert(5 <= cards && cards <= 30, "cube.cards range must be between 5 and 30");
   assert(typeof packs === "number", "cube.packs must be a number");
   assert(1 <= packs && packs <= 12, "cube.packs range must be between 1 and 12");
+  assert(0 <= burnsPerPack && burnsPerPack <= 4, "cube.burnsPerPack range must be between 0 and 2");
```
Comment (ZeldaZach) [inline backend/util.js:21]: Your string says "between 0 and 2" but your assertion is 0 and 4

**Comment by mixmix** [comment]
have started work over here on UI changes: https://github.com/dr4fters/dr4ft/pull/1351
it's gonna take me a couple days I think - summer holiday here in New Zealand and have kids running around

File: backend/util.js (hunk 1 above)
Comment (HerveH44) [reply backend/util.js]: Thanks for the review. I just corrected it.

**Comment by HerveH44** [comment]
@LotusWalker , @ZeldaZach , @tooomm , @mixmix. My part on this PR is implemented, the backend should be able to handle the burns for cube draft. You can try it on the branch deployed and let me know what you think. The UI is still underwork but we need your feedbacks!

**Comment by mixmix** [comment]
Ui just about done!

**Comment by mixmix** [comment]
BTW found a bug (I think) - you can pick and burn the same card...

**Comment by LotusWalker** [comment]
@HerveH44 Many thanks again for all your hard work on this. I have a 360 card cube and want to Glimpse draft with 2 friends, so our typical configuration is: 24 total packs (8 packs each) with 15 cards/pack, and this yields a total draft pool of 40 cards/player. Plenty enough for deckbuilding. All 360 cards in the cube are seen, with 120 drafted and 240 burned.

So I just attempted a draft on the branch deployed with the following settings: 
Number of Packs 8
Cards per Pack 15
Picks per Pack 1
Burns per Pack 2

I received the error message: "The cube needs between 960 and 100 000 cards with this settings, it has 360"

How should I modify the settings to make this work? For each player, we keep picking from the same pack after the first player has picked and burned. So for my cube example, a pack would look like this:

Player 1: Pick 1 and Burn 2
Player 2: Pick 1 and Burn 2
Player 3: Pick 1 and Burn 2
Player 1: Pick 1 and Burn 2
Player 2: Pick 1 and Burn 2

Then all 15 cards from that pack have been resolved, with 5 total picked and 10 total burned. So I'm not sure what I should fill in for the "Picks per pack" and "Burns per pack" to make this work.

**Comment by mixmix** [comment]
My UI branch is now ready for review and hopefully merge:
https://github.com/dr4fters/dr4ft/pull/1351

see first comment description of what's included

**Comment by mixmix** [comment]
@LotusWalker I tried to reproduce what you showed and I think the mistake you made was selecting "8 players" - 8*8*15 = 960 cards. If you change it to 2 players it should work fine

**Comment by HerveH44** [comment]
@LotusWalker how many players have you set as maximum for the draft? Maybe you did not set 8 players? Anyway, thanks for testing, we need some tweaking for sure. Can you post a link with your list so I can test it too?

**Comment by mixmix** [comment]
@HerveH44 can see 2 bugs:
1. order of pick burn
    - e.g. burn, burn, pick and this is not detected as "ready" while pick, burn, burn is.
2. you can pick and burn the same card
    - there's currently no logic preventing this
    - the pack will move on even though i haven't picked the right distinct cards...

**Comment by mixmix** [comment]
[EmanV1.txt](https://github.com/dr4fters/dr4ft/files/5760099/EmanV1.txt)

Here's a cube I've been testing with. I've been copying that into the text field.

**Comment by HerveH44** [comment]
@mixmix Thanks for the bug report. I'll check them shortly.

**Comment by HerveH44** [comment]
@mixmix your first bug report is a consequence of how autopick works. When you autopick a card, it only "locks" the card, it does not pick it. So if you want to validate your pick + burn, you have to pick another time your pick. 

I don't know what should be the normal behaviour ...

**Comment by tooomm** [comment]
I can't start a cube draft in the most recent deployment. Hitting `Create` does nothing, no error either.
Tried with a CubeCobra ID and pasting the above linked list.

**Comment by HerveH44** [comment]
@tooomm the error is very problematic :  FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory. It looks like Heroku nodeJs may not be able to handle the amount of data that we need. Best investigate it after the PR

**Comment by HerveH44** [comment]
I made the file cards.json twice smaller. It seems to make it work. @tooomm can you test again?

**Comment by mixmix** [comment]
@HerveH44 my expectation is that for glimpse burn and pick clicks behave the same. This is because they visually look the same.

When I go pick, burn, burn, I experience the app immediately confirming my choice, so when I go burn, burn, pick I expect the same check to be made. When it doesn't happen I'm confused (as a user) and don't know how to fix it. I then try another burn and it goes through, but I am not sure what the final state is that I executed.

Paths I see to resolve this :
- A. In glimpse add a check after each pick/ burn up ae if you've done all the needed selections. Auto execute those selections if you have
- B. remove all auto execution from glimpse, so that pick, burn, burn is only selections. A "confirm" button above the pack becomes visible which executes the selections
- C. Resolve it later if people have a problem!

(i think the double pick /burn is much more important)

**Comment by mixmix** [comment]
Testing - tested deployment on my phone and it was initially not working. Create page and to get laggy trying to make a cube glimpse. So I made a journal zendikar draft, then came back and the cube draft worked fine (immediately responded to create and was interacting to let me pick + burn)

**Bug 3:**
- create game with settings - 8packs, 1pick, 2 burn (so 3 selections) 
- run through first pack and your last choice only has 2 cards (8 % 3 == 2)
- game doesn't know I'm "done" when I go "pick, burn". I have to wait for time to elapse them out progresses

Possible solutions :
- D. Detect that all cards are selected (even though haven't selected 3) and execute
- E. Add a "confirm" button which executes  selections (same as B)
- F. Only allow `packs % (picks + burns) == 0`

**Bug 4**
- I can select 100 players and this is the first option in the dropdown
- this seems like a mistake I never want to draft with that many players.
- may not be this PRs fault?

**Comment by LotusWalker** [comment]
@HerveH44 @mixmix  Yeah you were right, I had accidentally set it to 8 players instead of 3. Just finished testing the following formats on my Vintage Cube:  http://cubecobra.com/cube/list/guertinvintagecube

-3 Player Modified Glimpse Draft (Burn 2): 24 Packs x 15 Cards/Pack, 8 Packs/Player; Draft Pool = 40 Cards/Player
-2 Player Burn 4 Draft: 24 Packs x 15 Cards/Pack, 12 Packs/Player; Draft Pool = 36 Cards/Player
-2 Player Impulse Draft (Burn 3): 18 Packs x 20 Cards/Pack, 9 Packs/Player; Draft Pool = 45 Cards/Player

Some Feedback:
Overall this looks great for a first pass! The pick/burn icons are clear and the rules are simple and have a nice layout. Some issues we noticed are the same ones that @mixmix noted above.

1) Key Issue: Requirement for proper number of Picks & Burns does not seem to be enforced. This is especially relevant when a player times out and autopick takes over. If I have just selected one card to burn for example, it does not automatically burn an additional card, and then the next player gets a pack with the incorrect number of cards in it. Changing your pick/burn selections might also possibly cause issues, but I'm not as sure about this.

2) UI Problem with MFDC and Transform cards: After you have made a selection and the card is in "Main Deck," when you hover over a MDFC or Transform card the image was huge (and some flickered) - way bigger than the previews of normal cards
(Tested on Shatterskull Smashing, Turntimber Symbiosis, Jace, Vryn's Prodigy, and Kytheon, Hero of Akros)

3) Feature Request - Can we change "Junk" to "Burned Cards" so we can show a player which cards they have previously burned in a draft?

4) Also agree that a "confirm" button could be a great addition if it's not too much work.

Otherwise it looks great, thanks & Happy New Year!

**Comment by mixmix** [comment]
@LotusWalker (2), the transform card bug was likely my work. I'll get that fixed.

Re the junk being the same as the burn pile - that's not a good idea because people can move things from junk to their deck.

I think we should add features like this in a later pass - the branch is already quite big.

Thanks for the thorough review
🍍 🐠

**Comment by LotusWalker** [comment]
Just did a little more testing. Can confirm that sometimes when you make a pick and then change it to a different card, it won't let you move on to the next pick and hangs until the timer runs out.

**Comment by mixmix** [comment]
Here's a patch for the broken flip card previews @LotusWalker found:
https://github.com/dr4fters/dr4ft/pull/1363

**Comment by HerveH44** [comment]
@mixmix I'd really like a confirm button. Could you help me on that, Idk where to start ...

**Comment by HerveH44** [comment]
@LotusWalker about the Junk space, this area is still part of your deck so we can't really change its use. 
Why would you need to know the cards you burnt?

**Comment by HerveH44** [comment]
@LotusWalker I can't reproduce the bug of timeout, I did couple tests, letting my pack timeout with or without autopicks or burn and it works fine. Also I fixed the bug with pack cards being all chosen. Remember sometimes you have to confirm your picks to tell server to pass the pack

**Comment by LotusWalker** [comment]
@HerveH44 Yeah, that makes sense. I like keeping track of burned cards so that I can remember them several packs later, but this is just personal preference, not a necessity.

**Comment by LotusWalker** [comment]
Re: Timeout issue: Yeah it was intermittent, and only seemed to happen when one of the drafters picked/burned a card and then changed it to something else. But perhaps I’m wrong, happy to test more tomorrow. 

Agree that a confirm button will eliminate all possible confusion and make it easy to lock in picks/burns.

**Comment by mixmix** [comment]
@HerveH44 with a confirm button ito code for it to work I imagine :
1. remove auto pass / confirm from the burn handler
2. add a computed value isCompleteSelection which is true when you have marked enough pick and burn cards
3. use that to enable a button which when clicked triggers the execution of selections and the passing of the pack

I don't know the mechanics of (3), i haven't read that logic yet.
 I notice we have a mixture of approaches to signaling in this branch though :
- App.emit from CardDefault.jsx
- directly calling functions in gamestate.js from CardGlimpse.jsx to pick/burn?

Would be nice to just do one style and keep things consistent if possible. (makes code easier to read / locate/ update)

Is that what you meant by knowing where to start?

**Comment by LotusWalker** [comment]
@HerveH44 @mixmix ok so I just had some time to do a bit more testing on the latest beta deployment with the following settings: 3 Player Modified Glimpse Draft (Burn 2): 24 Packs x 15 Cards/Pack, 8 Packs/Player; Draft Pool = 40 Cards/Player

1) We purposely tried to "break" the draft by skipping picks, skipping burns, timing out, changing picks frequently, etc. The main problem is that the draft is not enforcing the "1 Pick/2 Burn" rule for each selection. If someone times out, or just burns a card but doesn't pick one, etc. the system doesn't force the proper number of picks/burns to be made. Because of this, I ended up with a 28 card draft pool, while my friends each had over 40 cards (because the subsequent packs were distorted by my non-picks). When the picks/burns are enforced for each selection, each of the 3 players should end up with exactly 40 cards in their pool.

If someone doesn't make a pick or burn, ideally the system would randomly pick for them, and then randomly burn (so that every potential card was available for the auto-pick). Confirm button will also add a lot of clarity to the interface.

2) UI Issue: Some drafted cards didn't display properly in the "Main Deck." Took a couple screenshots of the issue:

https://imgur.com/a/S1BGbWt

**Comment by mixmix** [comment]
(2) is on me. I made it so each card has its text under it (nicely formatted) so you can always read in when images is loading slowly.

You'll notice the problem exists only for cards with super long text! I'll do some styling and come up with a fix

**Comment by mixmix** [comment]
@LotusWalker I suppose you tested after @HerveH44 latest commits? I think in future post "testing ui now" in feed so we can easily see which deploy you tested.

**Comment by LotusWalker** [comment]
@mixmix Yes exactly I used the latest deployment from 2 hours ago. Will def leave a quick notification message for the next round of tests.

**Comment by HerveH44** [comment]
Hey @LotusWalker , thanks for testing it ! It's a pleasure to have quick feedback on our work. I'll have a look at the issues you mentioned. A quick question tough: have you used bots? How was your selection when you left it go on timeout?

**Comment by LotusWalker** [comment]
@HerveH44 Happy to help! No, there were no bots at all in the tests we ran, just 3 human drafters. One reason I’m so excited about Glimpse drafts is bc we hate how the bots randomly pick cards...it leads to an unnatural draft where all the power wheels. Glimpse will be a whole lot more fun.

So we were doing a lot of clicking in an attempt to break things. It worked fine the first few rounds if you just pick, burn, burn, double click to move to the next pick. But problems arose once we started to change picks & burns or not make a pick/burn at all. Sometimes I would just wait until the timer ran out and do nothing, and I wouldn’t get a card added to my deck. Same if I burned 1 card but didn’t make a pick or burn the 2nd card.

Happy to test more tmr as needed. Cheers!

**Comment by mixmix** [comment]
Here's the  pr fixing the text overflow styling
https://github.com/dr4fters/dr4ft/pull/1369

File: frontend/src/gamestate.js [hunk 2]
```diff
@@ -187,27 +191,84 @@ class GameState extends EventEmitter {
...
+
+    if (packLength === (this.#autopickCardIds.length + this.#burnCardIds.length)) {
+      return true;
+    }
+
+    if (picksPerPack != this.#autopickCardIds.length) {
+      return false;
+    }
+
+    if (burnsPerPack != this.#burnCardIds.length) {
```
Comment (mixmix) [inline frontend/src/gamestate.js:267]: Recommend using `! ==` in case for some reason an `undefined` gets compared to a 0 because of a programmer error in future

**Comment by mixmix** [comment]
Here's an (optional) PR which renames some variables
https://github.com/dr4fters/dr4ft/pull/1371

**Comment by mixmix** [comment]
Here's a tiny bug fix for the bot burning
https://github.com/dr4fters/dr4ft/pull/1372

File: backend/human.js [hunk 3]
```diff
@@ -106,37 +102,72 @@ module.exports = class extends Player {
...
+    if (this.autopicks.length < min) {
+      // Remove autopick cards from selection
+      pullAt(pack, this.autopicks);
+
+      const remainingCardsToPick = min - this.autopicks.length;
+      times(remainingCardsToPick, () => {
+        const randomCard = sample(pack);
+        this.autopicks.push(randomCard.cardId);
+        pull(pack, randomCard);
+      });
```
[the remark this answers is not in the document]
Comment (mixmix) [reply backend/human.js:168]: See pr below exploring this

**Comment by mixmix** [comment]
Extended that tiny bug fix to catch the case where picks/ burns are topped up.
Is almost working but needs more work.

Burns seem to be topped up correctly. Picks do not seem to be.

**Comment by mixmix** [comment]
This bug fix branch is ready now
https://github.com/dr4fters/dr4ft/pull/1372

File: backend/human.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (mixmix) [reply backend/human.js:168]: MERGED!

**Comment by ZeldaZach** [comment]
@mixmix If you want to fixup this branch and get it ready, I'm curious what's left to do here :)

**Comment by mixmix** [comment]
The only thing I personally think left after that bug branch is merged is "add confirm button". I intend to do that, should be easy.

I've also got a branch where I make the icons and ux nicer

**Comment by mixmix** [comment]
I've played a glimpse draft with a friend (with no timer) and this swirled well.

There are some edge cases that need people to test manually - eg
8 packs, 1 pick 2 burn, 2 players, 1 bot, medium timer.
Then let the turns time out with different configurations of picked / burned selected

When we have redux in writing unit tests around this logic will be much easier

**Comment by mixmix** [comment]
@ZeldaZach here's a PR for the UX tidy. Also caught a bug and merged `master` in to resolve this current merge conflict
https://github.com/dr4fters/dr4ft/pull/1382#pullrequestreview-566829745

**Comment by mixmix** [comment]
@ZeldaZach here's the confirm button - one of the ways to confirm your selections for pick + burn
https://github.com/dr4fters/dr4ft/pull/1383

**Comment by LotusWalker** [comment]
@mixmix @HerveH44 Just did some testing on the latest beta deployment. Great progress guys, really like the addition of the "confirm" button, and we noticed no problems changing our picks this time around. However, we did find a couple bugs which I'll detail below. In general, if each drafter does the right thing...makes a pick followed by the correct number of burns, then clicks confirm, the draft has no issues at all. The problems arise when it times out, or when someone burns an incorrect number of cards.

Test Parameters: 2 Player (human) Impulse Draft (Pick 1, Burn 3): 20 cards/pack, 9 packs/player, Draft Pool = 45 cards/player

1) Graphical issue displaying cards: Sometimes, in both the draft area and the main deck area, the background text of a card will show through the card image. Including some screenshots below showing what happened. 

https://imgur.com/a/FuPjfAT
https://imgur.com/a/Ltdb4lP
https://imgur.com/a/OAaBp5E

Final Deck:  https://imgur.com/a/5SsUmvC

2) So, we again tried to "break" the draft. We noticed that if we just pick a card and let it time out, it burns the correct number of cards (3). However, if we pick 1 card and burn 1 card and let it time out, it burns only 2 cards total, not 3. So this causes that remaining card that should have been burned to go to the next player. That gives players an incorrect number of total cards. We each tried doing this several times throughout the draft, and I ended up with a 46 card draft pool, while my friend ended up with a 48 card draft pool. We should each have exactly 45 cards. I think the same error happens if you burn 2 cards, but then don't make a pick and don't burn a 3rd card.

Thanks again for all your great efforts, this is definitely almost there!

**Comment by mixmix** [comment]
Thanks so much @LotusWalker. This is invaluable <3

**Comment by mixmix** [comment]
Fixed **_1. Graphical issue displaying cards_** in https://github.com/dr4fters/dr4ft/pull/1382

**Comment by mixmix** [comment]
**2.  "break" the draft** still needs replicating and fixing. 

Note that I tried that setup previosly for pick 1 + burn 2 and it was behaving well. @HerveH44 let me know if you pick this bug up, otherwise I'll have another look when I have time

**Comment by LotusWalker** [comment]
@mixmix @HerveH44  Hey guys, just ran a test with two friends using the latest deployment. Unfortunately, we all were in agreement that the new user interface for picking and burning is much more confusing than the version we tested yesterday. What happened to the green "pick" checkmark and the red "burn" flame!? Those were great. The red and green outlines are much less clear, and "clicking" still doesn't really work. When you click a few times on a card or two, it seems totally random whether it's going to be picked or burned. So you have to keep clicking a bunch of times until it gets to where you want it to be. Can we bring back the red flame & green check mark? Thx for all your efforts!

**Comment by mixmix** [comment]
Tell me exactly what browser you're using? The flame and check should still be there in the bottom middle.
Also, worried about this clicking thing being weird. Have not seen anything like that

**Comment by LotusWalker** [comment]
So I am using Chrome, but neither myself or my two friends saw the flame or checkmark. I'm not sure what browsers they are using but I could find out for you.

**Comment by mixmix** [comment]
Thanks, and can you send the link of a deploy where you saw this?

**Comment by LotusWalker** [comment]
Ok just checked....2 of us were using Chrome and 1 was using Firefox. Perhaps the clicking randomness was because since we couldn't see the flame and checkmark we were just clicking random spots and were actually clicking the hidden keep/burn areas.

We tested using this link:  https://beta-dr4ft-glimpse-draf-wnjww9.herokuapp.com/

**Comment by mixmix** [comment]
How strange. Will look at this tonight.

**Comment by mixmix** [comment]
@LotusWalker I think the problem we were seeing was I was pulling in a new icon set using http and some browsers are blocking that now.

PR which should fix this: https://github.com/dr4fters/dr4ft/pull/1387

**Comment by mixmix** [comment]
I'm writing unit tests for the selection topper upper. I'm gonna make it a pure function (no side effects)

**Comment by HerveH44** [comment]
Sorry @mixmix , there's a pb with low qulity images : 
image

**Comment by HerveH44** [comment]
@mixmix  @LotusWalker , I think I found the problem for picks/burn on timeout. We were not calculating the right amount of cards. It was not visible until the last cards to pick/burn that's why @mixmix could not reproduce it. But when you test against a small pack it always showed up. It was because how we calculated the amount of cards to autopick/autoburn were not working with small numbers. 

@LotusWalker you can test it again when you can, you should not see this problem.

**Comment by HerveH44** [comment]
@mixmix: I don't know if the colors are voluntary but I think it's not clear at all whether cards are selected as picks or burn. In my screen, I see only red surroundings, the only info is the little flame or checkbox. Is it possible to have the "red" or "green" color surroundings ? 
As you can imagine I'm not very talented in CSS / frontend. I'll take a look but with limited time I may not solve this without your help.
image

**Comment by mixmix** [comment]
@HerveH44 cab you send a screen shot of the 'broken' state? I don't understand what you mean, and I am developing in Firefox so very surprised if something broken there.

Also did you push your fix for count. I don't see the commits in unis GitHub history

**Comment by mixmix** [comment]
Will fix the low quality images. I know what caused that

**Comment by HerveH44** [comment]
@mixmix the state is not broken, it's just that Firefox displays the wrong color, but from what I see it understand that's a "pick".

**Comment by mixmix** [comment]
Oh I see. Very weird. I haven't seen that in my testing. Sigh! Hopefully it's a typo

**Comment by mixmix** [comment]
Can't reproduce in firefox @HerveH44 (on current deploy on https://beta-dr4ft-glimpse-draf-wnjww9.herokuapp.com)
image

Here's what the DOM looks like for that selected pick:
image

can you confirm that you see `Selection -pick`.

**Comment by HerveH44** [comment]
image

**Comment by HerveH44** [comment]
Firefox version : 84.0.2 on OSx Catalina 10.15.7

**Comment by HerveH44** [comment]
Ok, I found the problem which is super weird: it's my extension "dark reader" that change the colors. It works fine if I don't use the extension!

**Comment by mixmix** [comment]
Lol. Yeah I saw that dark background and this crossed my mind and then I was like "nahhh, surely not"

**Comment by mixmix** [comment]
Here's the fix for low res images: https://github.com/dr4fters/dr4ft/pull/1388

**Comment by LotusWalker** [comment]
@mixmix @HerveH44 Just tested this out and everything worked flawlessly. The graphics are sleek and very clear, the pick/burn logic is enforced perfectly, and we saw no problems. Looks like this is now ready to go, great job guys!

**Comment by HerveH44** [comment]
Well done all, thanks @mixmix for all the amazing work and @LotusWalker for your super feedbacks. 

@ZeldaZach you can merge and deploy when you want.

**Comment by ZeldaZach** [comment]
I trust your judgement, merging.

**Comment by mixmix** [comment]
Woah, finally!
This was a long running PR with lots of side PRs.
Was really great getting timely feedback on code, design, and functionality.

Thanks in particular @LotusWalker to you and your crew of testers - you caught so many little bugs and presented them clearly, while also cheering us on ♥

I'm looking forward to working on some new PRs soon!

**Comment by ZeldaZach** [comment]
Please do join us on Discord at some point (https://mtgjson.com/discord) as I'd love to keep that #dr4ft room active 😄

**Comment by serokellcao** [comment]
Thank you! And sorry for not following through with my PR. I didn't have time until now to follow up on it.
Great job y'all!
